internal/cephfs/store/volumegroup.go

```diff
@@ -0,0 +1,285 @@
+*/
+
+package store
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ceph/ceph-csi/internal/cephfs/core"
+	cerrors "github.com/ceph/ceph-csi/internal/cephfs/errors"
+	fsutil "github.com/ceph/ceph-csi/internal/cephfs/util"
+	"github.com/ceph/ceph-csi/internal/util"
+	"github.com/ceph/ceph-csi/internal/util/log"
+
+	"github.com/container-storage-interface/spec/lib/go/csi"
+)
+
+type VolumeGroupOptions struct {
+	*VolumeOptions
+}
+
+// NewVolumeGroupOptions generates a new instance of volumeGroupOptions from the provided
+// CSI request parameters.
+func NewVolumeGroupOptions(
+	ctx context.Context,
+	req *csi.CreateVolumeGroupSnapshotRequest,
+	cr *util.Credentials,
+) (*VolumeGroupOptions, error) {
+	var (
+		opts = &VolumeGroupOptions{}
+		err  error
+	)
+
+	volOptions := req.GetParameters()
+	opts.VolumeOptions, err = getVolumeOptions(volOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = extractOptionalOption(&opts.NamePrefix, "volumeGroupNamePrefix", volOptions); err != nil {
+		return nil, err
+	}
+
+	opts.RequestName = req.GetName()
+
+	err = opts.Connect(cr)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err != nil {
+			opts.Destroy()
+		}
+	}()
+
+	fs := core.NewFileSystem(opts.conn)
+	opts.FscID, err = fs.GetFscID(ctx, opts.FsName)
+	if err != nil {
+		return nil, err
+	}
+
+	opts.MetadataPool, err = fs.GetMetadataPool(ctx, opts.FsName)
+	if err != nil {
+		return nil, err
+	}
+
+	return opts, nil
+}
+
+type VolumeGroupSnapshotIdentifier struct {
+	ReservedID                string
+	FsVolumeGroupSnapshotName string
+	VolumeGroupSnapshotID     string
+	RequestName               string
+	VolumeSnapshotMap         map[string]string
+}
+
+// GetVolumeIDs returns the list of volumeIDs in the VolumeSnaphotMap.
+func (vgsi *VolumeGroupSnapshotIdentifier) GetVolumeIDs() []string {
+	keys := make([]string, 0, len(vgsi.VolumeSnapshotMap))
+	for k := range vgsi.VolumeSnapshotMap {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+// NewVolumeGroupOptionsFromID generates a new instance of volumeGroupOptions and GroupIdentifier
+// from the provided CSI volumeGroupSnapshotID.
+func NewVolumeGroupOptionsFromID(
+	ctx context.Context,
+	volumeGroupSnapshotID string,
+	cr *util.Credentials,
+) (*VolumeGroupOptions, *VolumeGroupSnapshotIdentifier, error) {
+	var (
+		vi         util.CSIIdentifier
+		volOptions = &VolumeGroupOptions{}
+		vgs        VolumeGroupSnapshotIdentifier
+	)
+	// Decode the snapID first, to detect pre-provisioned snapshot before other errors
+	err := vi.DecomposeCSIID(volumeGroupSnapshotID)
+	if err != nil {
+		return volOptions, &vgs, cerrors.ErrInvalidVolID
+	}
+	volOptions.VolumeOptions = &VolumeOptions{}
+	volOptions.ClusterID = vi.ClusterID
+	vgs.VolumeGroupSnapshotID = volumeGroupSnapshotID
+	volOptions.FscID = vi.LocationID
+	vgs.ReservedID = vi.ObjectUUID
+
+	if volOptions.Monitors, err = util.Mons(util.CsiConfigFile, vi.ClusterID); err != nil {
+		return volOptions, &vgs, fmt.Errorf(
```

Comment:
Are the returned `volOptions` and `vgs` useful for the caller in case of an error? I prefer to return `nil` and not return incomplete/broken structs.